File: engine/rules/damage.py
"""Damage formula parsing and rolling."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Dict, Iterable

from engine.core.rng import DeterministicRNG


_DAMAGE_RE = re.compile(r"^(\d+)d(\d+)([+-]\d+)?$")
_FLAT_RE = re.compile(r"^[+-]?\d+$")

_DAMAGE_TYPE_ALIASES = {
    "lightning": "electricity",
    "pierce": "piercing",
    "slash": "slashing",
    "bludgeon": "bludgeoning",
}
_PHYSICAL_TYPES = {"bludgeoning", "piercing", "slashing"}
_ENERGY_TYPES = {"acid", "cold", "electricity", "fire", "force", "sonic"}


@dataclass
class DamageRoll:
    formula: str
    total: int
    rolls: list[int]
    flat_modifier: int


@dataclass
class DamageAdjustment:
    raw_total: int
    applied_total: int
    damage_type: str | None
    immune: bool
    resistance_total: int
    weakness_total: int


@dataclass
class AppliedDamage:
    incoming_total: int
    absorbed_by_temp_hp: int
    hp_loss: int
    new_hp: int
    new_temp_hp: int


def parse_formula(formula: str) -> tuple[int, int, int]:
    text = formula.strip()
    match = _DAMAGE_RE.match(text)
    if match:
        dice_count = int(match.group(1))
        dice_size = int(match.group(2))
        modifier = int(match.group(3) or 0)
        return dice_count, dice_size, modifier
    if _FLAT_RE.match(text):
        return 0, 1, int(text)
    raise ValueError(f"Unsupported damage formula: {formula}")


def roll_damage(rng: DeterministicRNG, formula: str, multiplier: int = 1) -> DamageRoll:
    dice_count, dice_size, modifier = parse_formula(formula)
    rolls = [rng.randint(1, dice_size).value for _ in range(dice_count)]
    total = (sum(rolls) + modifier) * multiplier
    return DamageRoll(
        formula=formula,
        total=max(0, total),
        rolls=rolls,
        flat_modifier=modifier,
    )


def _normalized_damage_type(raw: str | None) -> str | None:
    normalized = (raw or "").strip().lower()
    if not normalized:
        return None
    return _DAMAGE_TYPE_ALIASES.get(normalized, normalized)


def _damage_type_tags(damage_type: str | None) -> set[str]:
    normalized = _normalized_damage_type(damage_type)
    if normalized is None:
        return set()
    tags = {normalized}
    if normalized in _PHYSICAL_TYPES:
        tags.add("physical")
    if normalized in _ENERGY_TYPES:
        tags.add("energy")
    return tags


def _highest_matching_modifier(modifiers: Dict[str, int], damage_tags: set[str]) -> int:
    best = 0
    for key, value in modifiers.items():
        k = _normalized_damage_type(str(key))
        if k == "all" or (k is not None and k in damage_tags):
            best = max(best, int(value))
    return max(0, best)


def apply_damage_modifiers(
    *,
    raw_total: int,
    damage_type: str | None,
    resistances: Dict[str, int],
    weaknesses: Dict[str, int],
    immunities: Iterable[str],
) -> DamageAdjustment:
    raw = max(0, int(raw_total))
    normalized_type = _normalized_damage_type(damage_type)
    damage_tags = _damage_type_tags(normalized_type)
    immunity_set = {_normalized_damage_type(str(x)) or "" for x in immunities}

    if raw == 0:
        return DamageAdjustment(
            raw_total=0,
            applied_total=0,
            damage_type=normalized_type,
            immune=False,
            resistance_total=0,
            weakness_total=0,
        )

    if "all" in immunity_set or any(tag in immunity_set for tag in damage_tags):
        return DamageAdjustment(
            raw_total=raw,
            applied_total=0,
            damage_type=normalized_type,
            immune=True,
            resistance_total=0,
            weakness_total=0,
        )

    # Use the strongest matching value for each side, rather than stacking.
    resistance_total = _highest_matching_modifier(resistances, damage_tags)
    weakness_total = _highest_matching_modifier(weaknesses, damage_tags)

    applied = max(0, raw - resistance_total + weakness_total)
    return DamageAdjustment(
        raw_total=raw,
        applied_total=applied,
        damage_type=normalized_type,
        immune=False,
        resistance_total=max(0, resistance_total),
        weakness_total=max(0, weakness_total),
    )


def apply_damage_to_pool(*, hp: int, temp_hp: int, damage_total: int) -> AppliedDamage:
    incoming = max(0, int(damage_total))
    current_hp = max(0, int(hp))
    current_temp = max(0, int(temp_hp))
    absorbed = min(current_temp, incoming)
    hp_loss = max(0, incoming - absorbed)
    return AppliedDamage(
        incoming_total=incoming,
        absorbed_by_temp_hp=absorbed,
        hp_loss=hp_loss,
        new_hp=max(0, current_hp - hp_loss),
        new_temp_hp=max(0, current_temp - absorbed),
    )

File: src/rules/damage.ts
/**
 * Damage formula parsing and rolling.
 * Mirrors engine/rules/damage.py
 *
 * Pathfinder 2e ORC damage rules: resistances reduce, weaknesses increase,
 * immunities negate. Temp HP absorbs first.
 */

import { DeterministicRNG } from "../engine/rng";

const DAMAGE_RE = /^(\d+)d(\d+)([+-]\d+)?$/;
const FLAT_RE = /^[+-]?\d+$/;

const DAMAGE_TYPE_ALIASES: Record<string, string> = {
  lightning: "electricity",
  pierce: "piercing",
  slash: "slashing",
  bludgeon: "bludgeoning",
};

const PHYSICAL_TYPES = new Set(["bludgeoning", "piercing", "slashing"]);
const ENERGY_TYPES = new Set(["acid", "cold", "electricity", "fire", "force", "sonic"]);

export interface DamageRoll {
  formula: string;
  total: number;
  rolls: number[];
  flatModifier: number;
}

export interface DamageAdjustment {
  rawTotal: number;
  appliedTotal: number;
  damageType: string | null;
  immune: boolean;
  resistanceTotal: number;
  weaknessTotal: number;
}

export interface AppliedDamage {
  incomingTotal: number;
  absorbedByTempHp: number;
  hpLoss: number;
  newHp: number;
  newTempHp: number;
}

export function parseFormula(formula: string): [number, number, number] {
  const text = formula.trim();
  const match = DAMAGE_RE.exec(text);
  if (match) {
    const diceCount = parseInt(match[1], 10);
    const diceSize = parseInt(match[2], 10);
    const modifier = match[3] ? parseInt(match[3], 10) : 0;
    return [diceCount, diceSize, modifier];
  }
  if (FLAT_RE.test(text)) {
    return [0, 1, parseInt(text, 10)];
  }
  throw new Error(`Unsupported damage formula: ${formula}`);
}

export function rollDamage(
  rng: DeterministicRNG,
  formula: string,
  multiplier = 1,
): DamageRoll {
  const [diceCount, diceSize, modifier] = parseFormula(formula);
  const rolls: number[] = [];
  for (let i = 0; i < diceCount; i++) {
    rolls.push(rng.randint(1, diceSize).value);
  }
  const total = (rolls.reduce((s, r) => s + r, 0) + modifier) * multiplier;
  return {
    formula,
    total: Math.max(0, total),
    rolls,
    flatModifier: modifier,
  };
}

function normalizedDamageType(raw: string | null | undefined): string | null {
  const normalized = (raw ?? "").trim().toLowerCase();
  if (!normalized) return null;
  return DAMAGE_TYPE_ALIASES[normalized] ?? normalized;
}

function damageTypeTags(damageType: string | null): Set<string> {
  const normalized = normalizedDamageType(damageType);
  if (normalized === null) return new Set();
  const tags = new Set([normalized]);
  if (PHYSICAL_TYPES.has(normalized)) tags.add("physical");
  if (ENERGY_TYPES.has(normalized)) tags.add("energy");
  return tags;
}

function highestMatchingModifier(
  modifiers: Record<string, number>,
  damageTags: Set<string>,
  bypassTags: Set<string>,
): number {
  let best = 0;
  for (const [key, value] of Object.entries(modifiers)) {
    const k = normalizedDamageType(key);
    if (k === null) continue;
    if (bypassTags.has("all") || bypassTags.has(k)) continue;
    if (k === "all" || damageTags.has(k)) {
      best = Math.max(best, Number(value));
    }
  }
  return Math.max(0, best);
}

export function applyDamageModifiers(opts: {
  rawTotal: number;
  damageType?: string | null;
  resistances: Record<string, number>;
  weaknesses: Record<string, number>;
  immunities: string[];
  bypass?: string[] | null;
}): DamageAdjustment {
  const raw = Math.max(0, Math.floor(opts.rawTotal));
  const normalizedType = normalizedDamageType(opts.damageType);
  const damageTags = damageTypeTags(normalizedType);
  const immunitySet = new Set(
    (opts.immunities || []).map((x) => normalizedDamageType(String(x)) ?? ""),
  );
  const bypassSet = new Set(
    (opts.bypass || []).map((x) => normalizedDamageType(String(x)) ?? ""),
  );

  if (raw === 0) {
    return {
      rawTotal: 0,
      appliedTotal: 0,
      damageType: normalizedType,
      immune: false,
      resistanceTotal: 0,
      weaknessTotal: 0,
    };
  }

  // Check immunity
  if (
    (immunitySet.has("all") && !bypassSet.has("all")) ||
    [...damageTags].some((tag) => immunitySet.has(tag) && !bypassSet.has(tag))
  ) {
    return {
      rawTotal: raw,
      appliedTotal: 0,
      damageType: normalizedType,
      immune: true,
      resistanceTotal: 0,
      weaknessTotal: 0,
    };
  }

  const resistanceTotal = highestMatchingModifier(
    opts.resistances,
    damageTags,
    bypassSet,
  );
  const weaknessTotal = highestMatchingModifier(opts.weaknesses, damageTags, new Set());

  const applied = Math.max(0, raw - resistanceTotal + weaknessTotal);
  return {
    rawTotal: raw,
    appliedTotal: applied,
    damageType: normalizedType,
    immune: false,
    resistanceTotal: Math.max(0, resistanceTotal),
    weaknessTotal: Math.max(0, weaknessTotal),
  };
}

export function applyDamageToPool(opts: {
  hp: number;
  tempHp: number;
  damageTotal: number;
}): AppliedDamage {
  const incoming = Math.max(0, Math.floor(opts.damageTotal));
  const currentHp = Math.max(0, Math.floor(opts.hp));
  const currentTemp = Math.max(0, Math.floor(opts.tempHp));
  const absorbed = Math.min(currentTemp, incoming);
  const hpLoss = Math.max(0, incoming - absorbed);
  return {
    incomingTotal: incoming,
    absorbedByTempHp: absorbed,
    hpLoss,
    newHp: Math.max(0, currentHp - hpLoss),
    newTempHp: Math.max(0, currentTemp - absorbed),
  };
}
